speech_to_text/speech_to_text.livemd

```diff
@@ -0,0 +1,217 @@
+    {:ok, featurizer} = Bumblebee.load_featurizer({:hf, "openai/whisper-tiny"})
+    {:ok, tokenizer} = Bumblebee.load_tokenizer({:hf, "openai/whisper-tiny"})
+    {:ok, generation_config} = Bumblebee.load_generation_config({:hf, "openai/whisper-tiny"})
+
+    serving =
+      Bumblebee.Audio.speech_to_text_whisper(whisper, featurizer, tokenizer, generation_config,
+        defn_options: [compiler: EXLA]
+      )
+
+    Membrane.Logger.info("Whisper model ready")
+
+    state =
+      Map.merge(options, %{
+        serving: serving,
+        speech: <<>>,
+        queue: <<>>,
+        chunk_size: nil,
+        cut_silence: true
+      })
+
+    {[], state}
+  end
+
+  @impl true
+  def handle_stream_format(:input, stream_format, _ctx, state) do
+    chunk_size = Membrane.RawAudio.time_to_bytes(state.chunk_duration, stream_format)
+    {[stream_format: {:output, %Membrane.RemoteStream{}}], %{state | chunk_size: chunk_size}}
+  end
+
+  @impl true
+  def handle_buffer(:input, buffer, _ctx, state) do
+    input = state.queue <> buffer.payload
+
+    if byte_size(input) > @vad_chunk_size do
+      process_data(input, %{state | queue: <<>>})
+    else
+      {[], %{state | queue: input}}
+    end
+  end
+
+  defp process_data(data, state) do
+    # Here we filter out the silence at the beginning of each chunk.
+    # This way we can fit as much speech in a single chunk as possible
+    # and potentially remove whole silent chunks, which cause
+    # model hallucinations. If after removing the silence the chunk
+    # is not empty but too small to process, we store it in the state
+    # and prepend it to the subsequent chunk.
+    speech =
+      if state.speech == <<>> do
+        filter_silence(data, state)
+      else
+        state.speech <> data
+      end
+
+    if byte_size(speech) < state.chunk_size do
+      {[], %{state | speech: speech}}
+    else
+      model_input = Nx.from_binary(speech, :f32)
+      result = Nx.Serving.run(state.serving, model_input)
+      transcription = Enum.map_join(result.chunks, & &1.text)
+      buffer = %Membrane.Buffer{payload: transcription}
+      {[buffer: {:output, buffer}], %{state | speech: <<>>}}
+    end
+  end
+
+  defp filter_silence(samples, state) do
+    samples
+    |> generate_chunks()
+    |> Enum.drop_while(&(calc_volume(&1) < state.vad_threshold))
+    |> Enum.join()
+  end
+
+  defp generate_chunks(samples) when byte_size(samples) >= 2 * @vad_chunk_size do
+    <<chunk::binary-size(@vad_chunk_size), rest::binary>> = samples
+    [chunk | generate_chunks(rest)]
+  end
+
+  defp generate_chunks(samples) do
+    [samples]
+  end
+
+  defp calc_volume(chunk) do
+    samples = for <<sample::float-32-little <- chunk>>, do: sample
+    samples_cnt = Enum.count(samples)
+    samples_avg = Enum.sum(samples) / samples_cnt
+    sum_mean_square = samples |> Enum.map(&((&1 - samples_avg) ** 2)) |> Enum.sum()
+    :math.sqrt(sum_mean_square / samples_cnt)
```

Comment:
[NIT] We could probably delegate that part to some kind of `calculate_standard_deviation` function and describe why standard deviation represents volume

Reply:
Added a comment